### PR TITLE
Add Preact support

### PR DESCRIPTION
**Pull request by BartWaardenburg**
Issue:

Originally preact and react didn't differ much in their api's and building a storybook for preact components was as easy as including preact-compat to 'polyfill' the missing react api's for preact. Lately the pace at which react has been introducing new api's (forwardrefs, fragments, new context, etc.) has caused preact to lag behind a little bit and has caused the simple preact-compat solution to no longer work for building a preact-storybook.

With the popularity of both preact and storybook rising a dedicated and less error prone @storybook/preact package sound like the right direction to me.

## What I did

I added a new @storybook/preact package with custom preact renderer and a sample kitchen sink app to allow devs to build a custom storybook for preact. 

## How to test

I included a new kitchen sink app, but I don't know if I missed anything.

### CODE REVIEW

File: addons/centered/README.md [hunk 1]
```diff
@@ -47,6 +47,20 @@ storiesOf('MyComponent', module)
   }));
 ```
 
+example for Preact:
+
+```js
+import { storiesOf } from '@storybook/preact';
+import centered from '@storybook/addon-centered';
```
Comment (igor-dv) [inline addons/centered/README.md]:
```suggestion
import centered from '@storybook/addon-centered/preact';
```

File: examples/preact-kitchen-sink/package.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    "dev": "cross-env NODE_ENV=development webpack-dev-server --open --hot",
+    "storybook": "start-storybook -p 9009 -s public"
+  },
+  "dependencies": {
+    "global": "^4.3.2",
+    "preact": "^8.3.1"
+  },
+  "devDependencies": {
+    "@babel/core": "^7.1.2",
+    "@storybook/addon-actions": "4.1.0-alpha.10",
```
Comment (igor-dv) [inline examples/preact-kitchen-sink/package.json]: please use only supported addons here

File: examples/preact-kitchen-sink/src/stories/index.stories.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { h } from 'preact';
+
+import { storiesOf } from '@storybook/preact';
+import { action } from '@storybook/addon-actions';
+import { linkTo } from '@storybook/addon-links';
+
+import Button from '../Button';
+
+storiesOf('Welcome', module).add('to Storybook', () => (
+  <Button onclick={linkTo('Button')}>This button links to Button</Button>
```
Comment (igor-dv) [inline examples/preact-kitchen-sink/src/stories/index.stories.js]: The welcome story is a `<Button>` component 🤣 ?

File: examples/preact-kitchen-sink/jest.config.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+const config = require('../../jest.config');
```
Comment (igor-dv) [inline examples/preact-kitchen-sink/jest.config.js]: this file is needed only if you have tests (mostly for storyshots) in the example app

**Review by igor-dv — COMMENTED** [review]
Very cool 👍 

Things are missing for me:
* ADDONS_SUPPORT.md
* CLI for preact

File: examples/preact-kitchen-sink/src/stories/index.stories.js (hunk 3 above)
Comment (BartWaardenburg) [reply examples/preact-kitchen-sink/src/stories/index.stories.js]: Woops thanks, was still missing that one!

File: examples/preact-kitchen-sink/jest.config.js (hunk 4 above)
Comment (BartWaardenburg) [reply examples/preact-kitchen-sink/jest.config.js]: Thanks, I now added storyshots correctly to the example app.

File: examples/preact-kitchen-sink/package.json (hunk 2 above)
Comment (BartWaardenburg) [reply examples/preact-kitchen-sink/package.json]: The dependencies are updated to only display the supported addons.

File: docs/src/pages/basics/live-examples/index.md [hunk 5]
```diff
@@ -15,6 +15,7 @@ title: 'Live Examples'
 - [Svelte](https://storybooks-svelte.netlify.com/)
 - [Riot](https://storybooks-riot.netlify.com/)
 - [Ember](https://storybooks-ember.netlify.com/)
+- [Preact](https://storybooks-preact.netlify.com/)
```
[the remark this answers is not in the document]
Comment (igor-dv) [reply docs/src/pages/basics/live-examples/index.md]: We need to replace all these with `https://release-4-0--storybooks-{APP}.netlify.com/`

**Comment by igor-dv** [comment]
Preview: https://deploy-preview-4912--storybooks-preact.netlify.com 
I think the `welcome` story can be  more `preact`ish 
Take a look at:
https://deploy-preview-4912--storybooks-svelte.netlify.com/
https://deploy-preview-4912--storybooks-ember.netlify.com/

**Comment by igor-dv** [comment]
Also please add a symlink to the to the `preact-kitchen-sink` in the `examples/official-storybook/built-storybooks`

File: ADDONS_SUPPORT.md [hunk 6]
```diff
@@ -1,23 +1,23 @@
...
-|[jest](addons/jest)                         |+| | |+| | |+| | | | |
-|[knobs](addons/knobs)                       |+|+*|+|+|+|+|+|+|+|+|+|
-|[links](addons/links)                       |+|+|+|+|+|+|+| |+|+|+|
-|[notes](addons/notes)                       |+|+*|+|+|+|+|+| |+|+|+|
-|[options](addons/options)                   |+|+|+|+|+|+|+| |+|+|+|
-|[cssresources](addons/cssresources)         |+| |+|+|+|+|+|+|+|+|+|
-|[storyshots](addons/storyshots)             |+|+|+|+| | |+| |+|+| |
-|[storysource](addons/storysource)           |+| |+|+|+|+|+|+|+|+|+|
-|[viewport](addons/viewport)                 |+| |+|+|+|+|+|+|+|+|+|
+| | [React](app/react)|[React Native](app/react-native)|[Vue](app/vue)|[Angular](app/angular)| [Polymer](app/polymer)| [Preact](app/preact)| [Mithril](app/mithril)| [HTML](app/html)| [Marko](app/marko)| [Svelte](app/svelte)| [Riot](app/riot)| [Ember](app/ember)|
```
[the remark this answers is not in the document]
Comment (BartWaardenburg) [reply ADDONS_SUPPORT.md]: I reordered it to be in chronological order.

File: examples/official-storybook/built-storybooks/preact-kitchen-sink [hunk 7]
```diff
@@ -0,0 +1 @@
+/Users/p279825/Sites/storybook/examples/preact-kitchen-sink/storybook-static
```
Comment (igor-dv) [inline examples/official-storybook/built-storybooks/preact-kitchen-sink]: Looks like an absolute path. Take a look at others

File: ADDONS_SUPPORT.md [hunk 8]
```diff
@@ -1,23 +1,23 @@
...
-|[jest](addons/jest)                         |+| | |+| | |+| | | | |
-|[knobs](addons/knobs)                       |+|+*|+|+|+|+|+|+|+|+|+|
-|[links](addons/links)                       |+|+|+|+|+|+|+| |+|+|+|
-|[notes](addons/notes)                       |+|+*|+|+|+|+|+| |+|+|+|
-|[options](addons/options)                   |+|+|+|+|+|+|+| |+|+|+|
-|[cssresources](addons/cssresources)         |+| |+|+|+|+|+|+|+|+|+|
-|[storyshots](addons/storyshots)             |+|+|+|+| | |+| |+|+| |
-|[storysource](addons/storysource)           |+| |+|+|+|+|+|+|+|+|+|
-|[viewport](addons/viewport)                 |+| |+|+|+|+|+|+|+|+|+|
+| | [Angular](app/angular)|[Ember](app/ember)|[HTML](app/html)|[Marko](app/marko)|[Mithril](app/mithril)|[Polymer](app/polymer)|[Preact](app/preact)|[React](app/react)|[React Native](app/react-native)|[Riot](app/riot)|[Svelte](app/svelte)|[Vue](app/vue)|
```
Comment (igor-dv) [inline ADDONS_SUPPORT.md]: It's alphabetical =) I mean chronological, is where preact that have been added last will be in the end =)

**Comment by ndelangen** [comment]
@tmeasday I think you were able to solve the integration story last time by creating some symlink?

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
@BartWaardenburg This is amazing! such a cool addition! @shilman can we get this in for 4.1?

**Comment by igor-dv** [comment]
Let's just get https://github.com/storybooks/storybook/pull/4912#discussion_r240151665 fixed and we are ok.

**Comment by libetl** [comment]
preact is obviously something to integrate with storybook. I am one adopter of the framework.

File: .github/autolabeler.yml [hunk 9]
```diff
@@ -14,6 +14,7 @@
 'addon: viewport': ["addons/viewport/**"]
 'app: angular': ["app/angular/**"]
 'app: polymer ': ["app/polymer/**"]
+'app: preact': ["app/preact/**"]
 'app: react-native': ["app/react-native/**"]
 'app: react': ["app/react/**"]
```
Comment (libetl) [inline .github/autolabeler.yml]: oops, I missed the app: riot one.

File: app/preact/src/client/preview/render.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
+/* eslint-disable-next-line no-unused-vars */
+import { h, render } from 'preact';
```
Comment (libetl) [inline app/preact/src/client/preview/render.js]: will we have the pragma h pollution everywhere in storybook ?
Just need to make sure

File: app/preact/src/client/preview/render.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    });
+    return;
+  }
+
+  render(null, rootElement, renderedStory);
+
+  showMain();
+
+  renderedStory = render(element, rootElement);
+}
```
Comment (libetl) [inline app/preact/src/client/preview/render.js]: can you write some jest tests in the app/preact project too ?

File: ADDONS_SUPPORT.md (hunk 8 above)
Comment (BartWaardenburg) [reply ADDONS_SUPPORT.md]: 🤦

**Review by igor-dv — APPROVED** [review]
Good to go once Chromatic is green

**Comment by tmeasday** [comment]
I think the issue here is the symlink got messed up somehow. I'm not sure what it means, but I see:

```
 ~/OSS/storybook/examples/official-storybook   BartWaardenburg-next  ls -la built-storybooks
total 8
drwxr-xr-x  13 tom  staff  416 11 Dec 13:21 .
drwxr-xr-x  20 tom  staff  640 11 Dec 13:21 ..
lrwxr-xr-x   1 tom  staff   34 28 Feb  2018 angular-cli -> ../../angular-cli/storybook-static
lrwxr-xr-x   1 tom  staff   39 28 Feb  2018 cra-kitchen-sink -> ../../cra-kitchen-sink/storybook-static
-rw-r--r--   1 tom  staff   33 18 Oct 11:49 ember-cli
lrwxr-xr-x   1 tom  staff   41  8 May  2018 html-kitchen-sink -> ../../html-kitchen-sink/storybook-static/
lrwxr-xr-x   1 tom  staff   32 29 May  2018 marko-cli -> ../../marko-cli/storybook-static
lrwxr-xr-x   1 tom  staff   43 25 Apr  2018 mithril-kitchen-sink -> ../../mithril-kitchen-sink/storybook-static
lrwxr-xr-x   1 tom  staff   34 28 Feb  2018 polymer-cli -> ../../polymer-cli/storybook-static
lrwxr-xr-x   1 tom  staff   43 11 Dec 13:21 preact-kitchen-sink -> ../../preact-kitchen-sink/storybook-static?
lrwxr-xr-x   1 tom  staff   40 11 Sep 10:12 riot-kitchen-sink -> ../../riot-kitchen-sink/storybook-static
lrwxr-xr-x   1 tom  staff   42 11 Sep 10:12 svelte-kitchen-sink -> ../../svelte-kitchen-sink/storybook-static
lrwxr-xr-x   1 tom  staff   39 28 Feb  2018 vue-kitchen-sink -> ../../vue-kitchen-sink/storybook-static
```

**Pull request merged**